evalml/tests/conftest.py

```diff
@@ -1088,3 +1094,217 @@ def _imbalanced_data_X_y(problem_type, categorical_columns, size):
+                score_return_value.
+            mock_fit_side_effect: Passed as the side_effect argument of the pipeline.fit patch. Takes precedence over mock_fit_return_value.
+            mock_fit_return_value: Passed as the return_value argument of the pipeline.fit patch.
+            predict_proba_return_value: Passed as the return_value argument of the pipeline.predict_proba patch.
+            optimize_threshold_return_value: Passed as the return value of BinaryClassificationObjective.optimize_threshold patch.
+        """
+        mock_fit = self._patch_method(
+            "fit", side_effect=mock_fit_side_effect, return_value=mock_fit_return_value
+        )
+        mock_score = self._patch_method(
+            "score", side_effect=mock_score_side_effect, return_value=score_return_value
+        )
+
+        # For simplicity, we will always mock predict_proba and _encode_targets even if the problem is not a
+        # classification problem. For regression problems, we'll mock BinaryClassificationPipeline but it doesn't
+        # matter which one we mock since those methods won't be called for regression.
+        pipeline_to_mock = self._pipeline_class
+        if is_regression(self.problem_type):
+            pipeline_to_mock = "evalml.pipelines.BinaryClassificationPipeline"
+
+        mock_encode_targets = self._patch_method(
+            "_encode_targets",
+            side_effect=lambda y: y,
+            return_value=None,
+            pipeline_class_str=pipeline_to_mock,
+        )
+        mock_predict_proba = self._patch_method(
+            "predict_proba",
+            side_effect=None,
+            return_value=predict_proba_return_value,
+            pipeline_class_str=pipeline_to_mock,
+        )
+
+        mock_optimize = patch(
+            "evalml.objectives.BinaryClassificationObjective.optimize_threshold",
+            return_value=optimize_threshold_return_value,
+        )
+
+        mock_tell = patch("evalml.tuners.skopt_tuner.Optimizer.tell")
+
+        # Reset the mocks from a previous computation so that ValueError can be properly raised if
+        # user tries to access mocks before leaving the context
+        self._reset_mocks()
+
+        # Unfortunately, in order to set the MagicMock instances as class attributes we need to use the
+        # `with ... ` syntax.
+        with mock_fit as fit, mock_score as score, mock_encode_targets as encode, mock_predict_proba as proba, mock_tell as tell, mock_optimize as optimize:
+            # Can think of `yield` as blocking this method until the computation finishes running
+            yield
+            self._mock_fit = fit
+            self._mock_tell = tell
+            self._mock_score = score
+            self._mock_encode_targets = encode
+            self._mock_predict_proba = proba
+            self._mock_optimize_threshold = optimize
+
+    def run_search(
```

Comment:
What is the value of using this method over running
```python
with self.test_context():
    automl.search()
```
?
It seems like the main value is that `automl._SLEEP_TIME` is getting set. That's pretty great; I'm on board with standardizing that.

I think one disadvantage of using this method is that for an outside reader, there's more cognitive load to go look up with `_AutoMLTestEnv.run_search` does and see it runs `AutoMLSearch.search`, rather than just seeing that right away and understanding what the test is doing.

Assuming I'm right about the value here, is there some way we can set that sleep time in `test_context`, delete `run_search` and have all the test code still call `AutoMLSearch.search` directly?

One idea for doing that:
* Change the way `AutoMLSearch._SLEEP_TIME` is defined and used in `AutoMLSearch` so that the usage of it in `AutoMLSearch.search` could be mocked
* Update `test_context` to mock `SLEEP_TIME` to your value, 1e-7
* Optionally, add a `sleep_time` argument to `test_context` alongside the other args, if there are tests which want to set that value to something else.

I guess an alternative would be to make the sleep time a private `AutoMLSearch` init param 🤷‍♂️

Thoughts?

Reply:
@dsherry I think this is a great point! Yes, originally `run_search` was added to set the sleep time and because I thought that `create context then run search` was such a common operation that it deserved its own short-hand but I agree that it's clearer if we create the context for every test!

I'm implementing the first bullet you suggested here.